shell/ash.c

```diff
@@ -14783,11 +14784,19 @@ spawn_forkshell(struct job *jp, struct forkshell *fs, int mode)
 	char buf[16];
 	const char *argv[] = { "sh", "--forkshell", NULL, NULL };
 	intptr_t ret;
+	HANDLE hCurrentProcess;
 
 	new = forkshell_prepare(fs);
 	sprintf(buf, "%p", new->hMapFile);
 	argv[2] = buf;
 	ret = mingw_spawn_proc(argv);
+	if (ret != -1) {
+		hCurrentProcess = GetCurrentProcess();
+		if (!DuplicateHandle(hCurrentProcess, (HANDLE)ret, hCurrentProcess,
+				(HANDLE*)&ret, 0, TRUE, DUPLICATE_SAME_ACCESS)) {
```

Comment:
Hmm. `ret` is of the type `int`, but `HANDLE` [is a `void *`](https://github.com/git-for-windows/git-sdk-64/blob/1d62a677909cc8569e018ad2748874035a3cf367/mingw64/x86_64-w64-mingw32/include/winnt.h#L408). In 64-bit builds, this will potentially clip the value.

Any idea how we can fix this?

Reply:
`ret` is of type `intptr_t`, which is always capable of holding a pointer value.

Reply:
Oh, I was fooled by thinking that `ret` is the return value, and the function starts like this:

```c
/* FIXME: should consider running forkparent() and forkchild() */
static int
spawn_forkshell(struct job *jp, struct forkshell *fs, int mode)
{
	struct forkshell *new;
```

And the reason your patch works is because it is _not_ used as the return value, but instead it is passed to `forkparent()`, which in turn stores it in the `ps_proc` attribute of the `procstat` struct, and that's where `waitpid_child()` can pick it up.

Maybe this could be put into the commit message so that other reviewers do not have to deduce all of this themselves?

Reply:
Yes, I have update the commit message.